src/server/services/pipeline-in-progress/processData.js

```diff
@@ -0,0 +1,169 @@
+				readingOK = false;
+				errMsg = 'The first reading must be dropped when dealing only with endTimestamps.';
+			}
+			if (readingOK && startTimestamp.isSameOrAfter(endTimestamp)) {
+				readingOK = false;
+				errMsg = 'The reading end time is not after the start time.';
+				if (onlyEndTime) {
+					errMsg += ' The start time came from the previous readings end time.';
+				}
+			}
+			if (readingOK) {
+				// Check that startTimestamp is not before the previous endTimestamp
+				if (onlyEndTime && endTimestamp.isSameOrBefore(prevReading.endTimestamp)) {
+					readingOK = false;
+					errMsg = 'The reading is not after the previous reading with only end time given so we must drop the reading.';
+				}
+				else if (startTimestamp.isBefore(prevReading.endTimestamp)) {
+					if (isCumulative) {
+						readingOK = false;
+						errMsg = 'The reading start time is before the previous endTime and cumulative so OED cannot use this reading.'
+					}
+					else if (!isFirst(prevReading.endTimestamp)) {
+						//Only treat this as a warning since the readings may be sent in a different order.
+						errMsg = 'The current reading startTime is not after the previous reading\'s end time. This is expected on the first reading ever.';
+					}
+				}
+			}
+			if (readingOK && Math.abs(startTimestamp.diff(prevReading.endTimestamp)) > Tgap) {
+				if (isCumulative) {
+					readingOK = false;
+					errMsg = 'The end of the previous reading is too far from the start of the next readings in cumulative data so drop the readings.';
+				}
+				else if (!isFirst(prevReading.endTimestamp)) {
+					//Only treat this as a warning since we expect a gap in the first ever reading.
+					errMsg = 'There is a gap in time between this reading and the previous reading.';
+				}
+			}
+			if (readingOK) {
+				//This reading can be used
+				if (isCumulative) {
+					meterReading1 = rows[index - 2 * readingRepetition][0]; // use this if regular
+					meterReading2 = rows[index - readingRepetition][0];
+					meterReading = meterReading2 - meterReading1; // use this ifCumulative
+				}
+				else {
+					meterReading = rows[index - readingRepetition][0];
+				}
+				// Reject negative readings
+				if (isCumulative && meterReading1 < 0) {
+					// If the value isCumulative and the first reading is negative reject unless cumulativeReset is true
+					log.error(`DETECTED A NEGATIVE VALUE WHILE HANDLING CUMULATIVE READINGS FROM METER ${meterID}, ` +
+						`ROW ${index - readingRepetition}. REJECTED ALL READINGS`);
+					return [];
+				}
+				// To handle cumulative readings that resets at midnight
+				if (meterReading < 0) {
+					// if meterReading is negative and cumulative check that the times fall within an acceptable reset range
+					if (handleCumulativeReset(cumulativeReset, resetStart, resetEnd, startTimestamp)) {
+						meterReading = meterReading2;
+					}
+					else {
+						//cumulativeReset is not expected so throw an error
```

Comment:
Ok thank you, I've updated this so it returns an empty array which is equivalent to processing no readings. I think this makes sense since if there are unexpected net negative meter readings that cannot be handled it would be better to not insert any into the db and let the user know so they can check their data.